nab/sweeper.py

```diff
@@ -0,0 +1,245 @@
+from collections import namedtuple
+import logging
+import math
+
+from nab.corpus import Corpus
+
+logger = logging.getLogger(__name__)
+AnomalyPoint = namedtuple("AnomalyPoint", ["timestamp", "anomalyScore", "sweepScore", "windowName"])
+ThresholdScore = namedtuple("ThresholdScore", ["threshold", "score", "tp", "tn", "fp", "fn", "total"])
+
+
+def sigmoid(x):
+  """Standard sigmoid function."""
+  return 1 / (1 + math.exp(-x))
+
+
+def scaledSigmoid(relativePositionInWindow):
+  """Return a scaled sigmoid function given a relative position within a
+  labeled window.  The function is computed as follows:
+
+  A relative position of -1.0 is the far left edge of the anomaly window and
+  corresponds to S = 2*sigmoid(5) - 1.0 = 0.98661.  This is the earliest to be
+  counted as a true positive.
+
+  A relative position of -0.5 is halfway into the anomaly window and
+  corresponds to S = 2*sigmoid(0.5*5) - 1.0 = 0.84828.
+
+  A relative position of 0.0 consists of the right edge of the window and
+  corresponds to S = 2*sigmoid(0) - 1 = 0.0.
+
+  Relative positions > 0 correspond to false positives increasingly far away
+  from the right edge of the window. A relative position of 1.0 is past the
+  right  edge of the  window and corresponds to a score of 2*sigmoid(-5) - 1.0 =
+  -0.98661.
+
+  @param  relativePositionInWindow (float)  A relative position
+                                            within a window calculated per the
+                                            rules above.
+
+  @return (float)
+  """
+  if relativePositionInWindow > 3.0:
+    # FP well behind window
+    return -1.0
+  else:
+    return 2*sigmoid(-5*relativePositionInWindow) - 1.0
+
+
+
+class Sweeper(object):
+
+
+  def __init__(self, probationPercent=0.15, costMatrix=None):
+    self.probationPercent = probationPercent
+
+    self.tpWeight = None
+    self.fpWeight = None
+    self.fnWeight = None
+
+    if costMatrix is not None:
+      self.setCostMatrix(costMatrix)
+
+
+  def setCostMatrix(self, costMatrix):
+    self.tpWeight = costMatrix["tpWeight"]
+    self.fpWeight = costMatrix["fpWeight"]
+    self.fnWeight = costMatrix["fnWeight"]
+
+
+  def _getProbationaryLength(self, numRows):
+    return min(math.floor(self.probationPercent * numRows), self.probationPercent * 5000)
+
+
+  def _prepAnomalyListForScoring(self, inputAnomalyList):
+    """Sort by anomaly score and filter all rows with 'probationary' window name"""
+    return sorted(
+      [x for x in inputAnomalyList if x.windowName != 'probationary'],
+      key=lambda x: x.anomalyScore,
+      reverse=True)
+
+
+  def _prepareScoreByThresholdParts(self, inputAnomalyList):
+    scoreParts = {"fp": 0}
+    for row in inputAnomalyList:
+      if row.windowName not in ('probationary', None):
+        scoreParts[row.windowName] = -self.fnWeight
+    return scoreParts
+
+
+  def calcSweepScore(self, timestamps, anomalyScores, windowLimits, dataSetName):
```

Comment:
Add function comment, include cardinality of anomaly scores. From one data set or many? What does it return?

Reply:
Function comment added.